File: typescript/src/integrations/anthropic.ts
/**
 * AgentLens — Drop-in wrapper for the Anthropic TypeScript SDK.
 *
 * @example
 * ```ts
 * import Anthropic from "@anthropic-ai/sdk";
 * import { Profiler } from "../profiler";
 * import { ProfiledAnthropic } from "./anthropic";
 *
 * const profiler = new Profiler("my-agent");
 * const client = new ProfiledAnthropic(new Anthropic({ apiKey: "..." }), profiler);
 *
 * const message = await client.messages.create({
 *   model: "claude-opus-4-5",
 *   max_tokens: 1024,
 *   messages: [{ role: "user", content: "Hello!" }],
 * });
 * ```
 */

import { Profiler, TokenUsage } from "../profiler";

type AnyFn = (...args: unknown[]) => unknown;

function extractUsage(response: unknown): Partial<TokenUsage> | null {
  if (!response || typeof response !== "object") return null;
  const r = response as Record<string, unknown>;
  if (r["usage"] && typeof r["usage"] === "object") {
    const u = r["usage"] as Record<string, unknown>;
    if ("input_tokens" in u) {
      const prompt = Number(u["input_tokens"]) || 0;
      const completion = Number(u["output_tokens"]) || 0;
      return {
        promptTokens: prompt,
        completionTokens: completion,
        totalTokens: prompt + completion,
      };
    }
  }
  return null;
}

class ProfiledMessages {
  constructor(
    private readonly inner: { create: AnyFn; stream?: AnyFn },
    private readonly profiler: Profiler,
  ) {}

  async create(params: Record<string, unknown>, opts?: unknown): Promise<unknown> {
    const model = String(params["model"] ?? "unknown");
    const name = `anthropic.messages/${model}`;
    const call = this.profiler.startCall(name, "llm", model);

    try {
      const response = await (this.inner.create as (p: unknown, o?: unknown) => Promise<unknown>)(
        params,
        opts,
      );
      const usage = extractUsage(response);
      const inferredModel =
        response && typeof response === "object" && "model" in (response as object)
          ? String((response as Record<string, unknown>)["model"])
          : model;
      this.profiler.endCall(call, {
        success: true,
        tokenUsage: usage ?? undefined,
        model: inferredModel,
      });
      return response;
    } catch (err) {
      this.profiler.endCall(call, {
        success: false,
        error: err instanceof Error ? err : new Error(String(err)),
      });
      throw err;
    }
  }

  /**
   * Wraps the streaming `client.messages.stream(...)` helper.
   */
  stream(params: Record<string, unknown>): unknown {
    const model = String(params["model"] ?? "unknown");
    const name = `anthropic.messages.stream/${model}`;
    const call = this.profiler.startCall(name, "llm", model);
    const profiler = this.profiler;

    if (!this.inner.stream) {
      throw new Error("Underlying Anthropic client does not support .stream()");
    }

    const streamCtx = (this.inner.stream as (p: unknown) => unknown)(params);

    // Wrap in a proxy that records the call when the stream finalizes
    return new Proxy(streamCtx as object, {
      get(target, prop) {
        const value = Reflect.get(target, prop);
        if (prop === "finalMessage" && typeof value === "function") {
          return async function (this: unknown, ...args: unknown[]) {
            try {
              const result = await (value as AnyFn).apply(target, args);
              const usage = extractUsage(result);
              profiler.endCall(call, { success: true, tokenUsage: usage ?? undefined });
              return result;
            } catch (err) {
              profiler.endCall(call, {
                success: false,
                error: err instanceof Error ? err : new Error(String(err)),
              });
              throw err;
            }
          };
        }
        if (typeof value === "function") {
          return value.bind(target);
        }
        return value;
      },
    });
  }
}

/**
 * Drop-in wrapper for an `Anthropic` client instance.
 */
export class ProfiledAnthropic {
  readonly messages: ProfiledMessages;

  constructor(
    private readonly client: {
      messages: { create: AnyFn; stream?: AnyFn };
      [key: string]: unknown;
    },
    public readonly profiler: Profiler = new Profiler("anthropic"),
  ) {
    this.messages = new ProfiledMessages(client.messages, profiler);
  }

  [key: string]: unknown;
}

File: python/agentlens/integrations/anthropic.py
"""
Drop-in wrapper for the Anthropic Python client.

Usage::

    import anthropic
    from agentlens import Profiler
    from agentlens.integrations.anthropic import ProfiledAnthropic

    profiler = Profiler("my-agent")
    client = ProfiledAnthropic(anthropic.Anthropic(api_key="..."), profiler=profiler)

    # Use exactly like the standard anthropic client
    message = client.messages.create(
        model="claude-opus-4-5",
        max_tokens=1024,
        messages=[{"role": "user", "content": "Hello!"}],
    )
"""

from __future__ import annotations

from typing import Any, Iterator, Optional

from ..profiler import CallType, ProfiledCall, Profiler, TokenUsage


class _ProfiledMessages:
    def __init__(self, inner: Any, profiler: Profiler) -> None:
        self._inner = inner
        self._profiler = profiler

    def create(self, *args: Any, **kwargs: Any) -> Any:
        model = kwargs.get("model", "unknown")
        stream = kwargs.get("stream", False)
        name = f"anthropic.messages/{model}"

        if stream:
            return self._create_stream(name, *args, **kwargs)

        call = self._profiler._start_call(name, CallType.LLM, model=model)
        try:
            response = self._inner.create(*args, **kwargs)
            usage = TokenUsage.from_anthropic_usage(getattr(response, "usage", None))
            inferred_model = getattr(response, "model", model)
            call.finish(success=True, token_usage=usage, model=inferred_model)
            return response
        except Exception as exc:
            call.finish(success=False, error=exc)
            raise
        finally:
            self._profiler._record(call)

    def _create_stream(self, name: str, *args: Any, **kwargs: Any) -> Any:
        model = kwargs.get("model", "unknown")
        call = self._profiler._start_call(name, CallType.LLM, model=model)
        try:
            stream = self._inner.create(*args, **kwargs)
        except Exception as exc:
            call.finish(success=False, error=exc)
            self._profiler._record(call)
            raise
        return _AnthropicStreamWrapper(stream, call, self._profiler)

    async def acreate(self, *args: Any, **kwargs: Any) -> Any:
        model = kwargs.get("model", "unknown")
        name = f"anthropic.messages/{model}"
        call = self._profiler._start_call(name, CallType.LLM, model=model)
        try:
            response = await self._inner.acreate(*args, **kwargs)
            usage = TokenUsage.from_anthropic_usage(getattr(response, "usage", None))
            call.finish(success=True, token_usage=usage, model=getattr(response, "model", model))
            return response
        except Exception as exc:
            call.finish(success=False, error=exc)
            raise
        finally:
            self._profiler._record(call)

    def stream(self, *args: Any, **kwargs: Any) -> Any:
        """Support the `with client.messages.stream(...)` context manager pattern."""
        model = kwargs.get("model", "unknown")
        name = f"anthropic.messages.stream/{model}"
        call = self._profiler._start_call(name, CallType.LLM, model=model)
        try:
            ctx = self._inner.stream(*args, **kwargs)
        except Exception as exc:
            call.finish(success=False, error=exc)
            self._profiler._record(call)
            raise
        return _AnthropicStreamContext(ctx, call, self._profiler)


class _AnthropicStreamWrapper:
    def __init__(self, stream: Any, call: ProfiledCall, profiler: Profiler) -> None:
        self._stream = stream
        self._call = call
        self._profiler = profiler

    def __iter__(self) -> Iterator[Any]:
        try:
            for event in self._stream:
                yield event
            self._call.finish(success=True)
        except Exception as exc:
            self._call.finish(success=False, error=exc)
            raise
        finally:
            self._profiler._record(self._call)

    def __getattr__(self, item: str) -> Any:
        return getattr(self._stream, item)


class _AnthropicStreamContext:
    """Wraps the context manager returned by `client.messages.stream(...)`."""

    def __init__(self, ctx: Any, call: ProfiledCall, profiler: Profiler) -> None:
        self._ctx = ctx
        self._call = call
        self._profiler = profiler

    def __enter__(self) -> Any:
        inner = self._ctx.__enter__()
        return _AnthropicStreamWrapper(inner, self._call, self._profiler)

    def __exit__(self, *args: Any) -> Any:
        return self._ctx.__exit__(*args)

    async def __aenter__(self) -> Any:
        inner = await self._ctx.__aenter__()
        return _AnthropicStreamWrapper(inner, self._call, self._profiler)

    async def __aexit__(self, *args: Any) -> Any:
        return await self._ctx.__aexit__(*args)


class ProfiledAnthropic:
    """
    Wraps an ``anthropic.Anthropic`` instance and automatically profiles
    every API call.

    All non-wrapped attributes are forwarded transparently to the underlying
    client so existing code requires no other changes.
    """

    def __init__(self, client: Any, profiler: Optional[Profiler] = None) -> None:
        self._client = client
        self._profiler = profiler or Profiler("anthropic")
        self.messages = _ProfiledMessages(client.messages, self._profiler)

    def __getattr__(self, item: str) -> Any:
        return getattr(self._client, item)

    @property
    def profiler(self) -> Profiler:
        return self._profiler
